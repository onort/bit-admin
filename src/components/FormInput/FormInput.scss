@import "../../styles/main.scss";

.container {
  width: 100%;
  margin: $spacer-sm 0;
}

.label {
  display: block;
  font-size: $font-size-md;
  font-weight: 600;
  letter-spacing: 0.66px;
  margin-bottom: $spacer-xxsm;
}

.required:after {
  content: " *";
  margin-left: px-to-rem(-2px);
}

.input {
  display: block;
  width: 100%;
  border: 1px solid $light-border;
  background-color: $white;
  border-radius: px-to-rem(4px);
  padding: $spacer-sm;
  margin-bottom: $spacer-xxsm;
}

.error {
  display: block;
  font-size: $font-size-sm;
  color: $danger;
}
