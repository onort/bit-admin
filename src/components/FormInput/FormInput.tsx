import React from "react"
import { FieldProps } from "formik"
import cx from "classnames"

import styles from "./FormInput.scss"
import { ErrorMessage } from "../"

// Custom component props are untyped as of formik v1.3.1
// half & required uses 0 and 1 instead of a boolean
// reason being prop passed through Formik Field component
interface Props extends FieldProps {
  half?: 0 | 1
  label?: string
  required?: 0 | 1
}

const FormInput: React.SFC<Props> = ({ form, field, ...props }) => {
  const { errors, touched } = form
  const { name } = field
  const containerClassName = cx(styles.container, {
    [styles.half]: !!props.half
  })
  const labelClassName = cx(styles.label, {
    [styles.required]: !!props.required
  })
  return (
    <div className={containerClassName}>
      {props.label && (
        <label className={labelClassName} htmlFor={name}>
          {props.label}
        </label>
      )}
      <input
        {...field}
        {...props}
        className={styles.input}
        required={!!props.required}
      />
      {touched[name] &&
        errors[name] && <ErrorMessage small={true} message={errors[name]} />}
    </div>
  )
}

FormInput.defaultProps = {
  half: 0,
  required: 0
}

export default FormInput
