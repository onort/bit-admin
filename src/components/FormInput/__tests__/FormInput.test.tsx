import React from "react"
import { mount, shallow } from "enzyme"
import { Formik, Form, Field } from "formik"

import FormInput from "../"

describe("<FormInput />", () => {
  it("should match snapshot", () => {
    const fakeSubmit = jest.fn()
    const wrapper = shallow(
      <Formik initialValues={{ test: "" }} onSubmit={fakeSubmit}>
        <Form>
          <Field type="text" name="test" label="Test" component={FormInput} />
        </Form>
      </Formik>
    )
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should render a label element with correct text", () => {
    const fakeSubmit = jest.fn()
    const testLabel = "Test"
    const wrapper = mount(
      <Formik initialValues={{ test: "" }} onSubmit={fakeSubmit}>
        <Form>
          <Field
            type="text"
            name="test"
            label={testLabel}
            component={FormInput}
          />
        </Form>
      </Formik>
    )
    expect(wrapper.find("label").exists()).toBe(true)
    expect(wrapper.find("label").text()).toBe(testLabel)
    wrapper.unmount()
  })

  it("should render correctly if half prop is passed", () => {
    const fakeSubmit = jest.fn()
    const wrapper = mount(
      <Formik initialValues={{ test: "" }} onSubmit={fakeSubmit}>
        <Form>
          <Field
            type="text"
            name="test"
            label="Test"
            half={1}
            component={FormInput}
          />
        </Form>
      </Formik>
    )
    console.log(wrapper.html())
    expect(wrapper.find(".container.half").exists()).toBe(true)
    wrapper.unmount()
  })

  it.skip("should error text when there's an error", () => {
    // Test fails, need to handle async validation?
    const fakeSubmit = jest.fn()
    const failValidate = () => ({ test: "Error" })
    const testLabel = "Test"
    const wrapper = mount(
      <Formik
        initialValues={{ test: "" }}
        onSubmit={fakeSubmit}
        validate={failValidate}
      >
        <Form>
          <Field
            type="text"
            name="test"
            label={testLabel}
            component={FormInput}
          />
          <button type="submit">Submit</button>
        </Form>
      </Formik>
    )
    expect(wrapper.find("button").simulate("click"))
    expect(fakeSubmit).toBeCalledTimes(0)
    expect(wrapper.find(".error").exists()).toBe(true)
    wrapper.unmount()
  })
})
