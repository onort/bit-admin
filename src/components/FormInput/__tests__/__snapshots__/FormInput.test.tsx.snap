// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormInput /> should match snapshot 1`] = `"<form><div class=\\"container\\"><label class=\\"label\\" for=\\"test\\">Test</label><input type=\\"text\\" value=\\"\\" name=\\"test\\" label=\\"Test\\" half=\\"0\\" class=\\"input\\"/></div></form>"`;
